File: java/base/src/main/vsdk/toolkit/common/dataStructures/CircularDoubleLinkedList.java
package vsdk.toolkit.common.dataStructures;

import java.io.Serial;
import vsdk.toolkit.common.FundamentalEntity;
import vsdk.toolkit.common.VSDK;
import vsdk.toolkit.common.logging.Logger;

public class CircularDoubleLinkedList<E> extends FundamentalEntity
{
    @Serial private static final long serialVersionUID = 20070422L;

    private _CircularDoubleLinkedListNode<E> head;
    private _CircularDoubleLinkedListNode<E> window;
    private int currentSize;

    // Sequential-access memo for get(int): the dominant access pattern in the
    // CSG kernel is `for (i = 0; i < list.size(); i++) list.get(i)`, which on a
    // plain head-relative scan is O(n^2) per loop. Caching the last accessed
    // (index, node) lets an ascending walk continue from where it left off,
    // making such loops O(n) amortized. Reset on any structural change. Marked
    // transient so it never participates in serialization/equality.
    private transient int accessMemoIndex;
    private transient _CircularDoubleLinkedListNode<E> accessMemoNode;

    public CircularDoubleLinkedList()
    {
        head = null;
        window = null;
        currentSize = 0;
        invalidateAccessMemo();
    }

    private void invalidateAccessMemo()
    {
        accessMemoIndex = -1;
        accessMemoNode = null;
    }

    public int size()
    {
        return currentSize;
    }

    public void add(E e)
    {
        _CircularDoubleLinkedListNode<E> newContainer;
        newContainer = new _CircularDoubleLinkedListNode<E>();
        newContainer.data = e;
        if ( head == null ) {
            head = newContainer;
            newContainer.next = newContainer;
            newContainer.previous = newContainer;
        }
        else {
            newContainer.previous = head.previous;
            newContainer.next = head;
            head.previous.next = newContainer;
            head.previous = newContainer;
        }
        currentSize++;
        invalidateAccessMemo();
    }

    public void insertBefore(E newElem, E pivot)
    {
        locateWindowAtElem(pivot);
        _CircularDoubleLinkedListNode<E> newContainer;

        newContainer = new _CircularDoubleLinkedListNode<E>();
        newContainer.data = newElem;

        if ( head == null ) {
            head = newContainer;
            newContainer.next = newContainer;
            newContainer.previous = newContainer;
        }
        else if ( window == null || window == head ) {
            window = head;
            head = newContainer;
            newContainer.previous = window.previous;
            newContainer.next = window;
            window.previous.next = newContainer;
            window.previous = newContainer;
        }
        else {
            newContainer.previous = window.previous;
            newContainer.next = window;
            window.previous.next = newContainer;
            window.previous = newContainer;
        }
        currentSize++;
        invalidateAccessMemo();
    }

    public void locateWindowAtIndex(int index)
    {
        if ( index < 0 || index >= currentSize ) {
            return;
        }
        int i;
        for ( i = 0, window = head;
              i < currentSize && i < index;
              i++, window = window.next ) {
            // Note this updates the window pointer
	}
    }

    public boolean locateWindowAtElem(E e)
    {
        int i;

        for ( i = 0, window = head;
              i < currentSize;
              i++, window = window.next ) {
            if ( window.data == e ) {
                return true;
            }
        }
        window = null;
        return false;
    }

    public void swapElements(E e1, E e2)
    {
        locateWindowAtElem(e1);
        _CircularDoubleLinkedListNode<E> window1 = window;
        locateWindowAtElem(e2);
        _CircularDoubleLinkedListNode<E> window2 = window;

        if ( window1 == null || window2 == null ) return;
        E temp = window1.data;

        window1.data = window2.data;
        window2.data = temp;
        invalidateAccessMemo();
    }

    public E next()
    {
        if ( window == null ) {
            window = head;
        }
        E elem = window.data;
        window = window.next;
        return elem;
    }

    public E getWindow()
    {
        if ( head == null ) return null;
        if ( window == null ) {
            window = head;
        }
        return window.data;
    }

    public E previous()
    {
        if ( window == null ) {
            window = head;
        }
        E elem = window.data;
        window = window.previous;
        return elem;
    }

    public E nextOf(E e)
    {
        _CircularDoubleLinkedListNode<E> current;
        int i;

        for ( i = 0, current = head;
              i < currentSize; i++, current = current.next ) {
            if ( current.data == e ) {
                return current.next.data;
            }
        }
        return null;
    }

    public E previousOf(E e)
    {
        _CircularDoubleLinkedListNode<E> current;
        int i;

        for ( i = 0, current = head;
              i < currentSize; i++, current = current.next ) {
            if ( current.data == e ) {
                return current.previous.data;
            }
        }
        return null;
    }

    public E get(int index)
    {
        _CircularDoubleLinkedListNode<E> current;

        if ( index < 0 || index >= currentSize ) {
            // Report index out of bounds exception!
            String msg;
            msg = "IndexOutOfBounds Exception! - Trying to `get` with index " + index + " in a list with " + currentSize + " elements.";

            Logger.reportMessage(this, VSDK.FATAL_ERROR, "get", msg);
            return null;
        }
        int i;
        // Start from the access memo when it is at or before the requested
        // index, so ascending get(i) walks continue forward instead of
        // restarting from head; otherwise scan from head.
        if ( accessMemoNode != null && accessMemoIndex >= 0 &&
             accessMemoIndex <= index ) {
            i = accessMemoIndex;
            current = accessMemoNode;
        }
        else {
            i = 0;
            current = head;
        }
        for ( ; i < currentSize && i < index; i++, current = current.next ) {
            // Local traversal keeps indexed reads independent from window.
	}
        accessMemoIndex = index;
        accessMemoNode = current;
        return current.data;
    }

    public void remove(int pos)
    {
        locateWindowAtIndex(pos);
        removeElemAtWindow();
    }

    public void removeElemAtWindow()
    {
        if ( window == null ) return;
        if ( window == head ) head = window.next;
        window.previous.next = window.next;
        window.next.previous = window.previous;
        window = null;
        currentSize--;
        invalidateAccessMemo();
    }

    public void push(E newElem)
    {
        window = head;
        _CircularDoubleLinkedListNode<E> newContainer;

        newContainer = new _CircularDoubleLinkedListNode<E>();
        newContainer.data = newElem;

        if ( head == null ) {
            head = newContainer;
            newContainer.next = newContainer;
            newContainer.previous = newContainer;
        }
        else if ( window == null || window == head ) {
            window = head;
            head = newContainer;
            newContainer.previous = window.previous;
            newContainer.next = window;
            window.previous.next = newContainer;
            window.previous = newContainer;
        }
        else {
            newContainer.previous = window.previous;
            newContainer.next = window;
            window.previous.next = newContainer;
            window.previous = newContainer;
        }
        currentSize++;
        invalidateAccessMemo();
    }

    public void reverse()
    {
        _CircularDoubleLinkedListNode<E> ptr, qtr;
        E tmp;
        int i = 0;

        ptr = head;
        qtr = head.previous;
        do {
            tmp = ptr.data;
            ptr.data = qtr.data;
            qtr.data = tmp;

            ptr = ptr.next;
            qtr = qtr.previous;
            i++;
        } while ( ptr != head && i < currentSize/2 );
        invalidateAccessMemo();
    }

}

File: typescript/base/src/main/vsdk/toolkit/common/dataStructures/CircularDoubleLinkedList.ts
import { FundamentalEntity } from "../FundamentalEntity.js";
import { VSDK } from "../VSDK.js";
import { Logger } from "../logging/Logger.js";
import { _CircularDoubleLinkedListNode } from "./_CircularDoubleLinkedListNode.js";

/** Mutable circular, double-linked list with the Java window-cursor contract. */
export class CircularDoubleLinkedList<E> extends FundamentalEntity {
    private head: _CircularDoubleLinkedListNode<E> | null = null;
    private window: _CircularDoubleLinkedListNode<E> | null = null;
    private currentSize = 0;
    private accessMemoIndex = -1;
    private accessMemoNode: _CircularDoubleLinkedListNode<E> | null = null;

    public size(): number {
        return this.currentSize;
    }
    public add(element: E): void {
        this.insertAtEnd(element);
    }
    public push(element: E): void {
        this.insertBeforeHead(element);
    }

    public insertBefore(newElement: E, pivot: E): void {
        this.locateWindowAtElem(pivot);
        if (this.head === null || this.window === null || this.window === this.head) this.insertBeforeHead(newElement);
        else this.insertBeforeNode(newElement, this.window);
    }

    public locateWindowAtIndex(index: number): void {
        if (!Number.isInteger(index) || index < 0 || index >= this.currentSize || this.head === null) return;
        let node = this.head;
        for (let i = 0; i < index; i++) node = node.next!;
        this.window = node;
    }

    public locateWindowAtElem(element: E): boolean {
        if (this.head === null) {
            this.window = null;
            return false;
        }
        let node = this.head;
        for (let i = 0; i < this.currentSize; i++, node = node.next!) {
            if (node.data === element) {
                this.window = node;
                return true;
            }
        }
        this.window = null;
        return false;
    }

    public swapElements(first: E, second: E): void {
        if (!this.locateWindowAtElem(first)) return;
        const firstNode = this.window!;
        if (!this.locateWindowAtElem(second)) return;
        const secondNode = this.window!;
        const data = firstNode.data;
        firstNode.data = secondNode.data;
        secondNode.data = data;
        this.invalidateAccessMemo();
    }

    public next(): E | null {
        if (this.window === null) this.window = this.head;
        if (this.window === null) return null;
        const data = this.window.data;
        this.window = this.window.next;
        return data;
    }
    public previous(): E | null {
        if (this.window === null) this.window = this.head;
        if (this.window === null) return null;
        const data = this.window.data;
        this.window = this.window.previous;
        return data;
    }
    public getWindow(): E | null {
        if (this.head === null) return null;
        if (this.window === null) this.window = this.head;
        return this.window.data;
    }

    public nextOf(element: E): E | null {
        const node = this.nodeFor(element);
        return node?.next?.data ?? null;
    }
    public previousOf(element: E): E | null {
        const node = this.nodeFor(element);
        return node?.previous?.data ?? null;
    }

    public get(index: number): E | null {
        if (!Number.isInteger(index) || index < 0 || index >= this.currentSize) {
            Logger.reportMessage(
                this,
                VSDK.FATAL_ERROR,
                "get",
                `IndexOutOfBounds Exception! - Trying to \`get\` with index ${index} in a list with ${this.currentSize} elements.`,
            );
            return null;
        }
        let i =
            this.accessMemoNode !== null && this.accessMemoIndex >= 0 && this.accessMemoIndex <= index
                ? this.accessMemoIndex
                : 0;
        let node =
            i === 0 && (this.accessMemoNode === null || this.accessMemoIndex > index)
                ? this.head!
                : this.accessMemoNode!;
        for (; i < index; i++) node = node.next!;
        this.accessMemoIndex = index;
        this.accessMemoNode = node;
        return node.data;
    }

    public remove(position: number): void {
        this.locateWindowAtIndex(position);
        this.removeElemAtWindow();
    }
    public removeElemAtWindow(): void {
        const node = this.window;
        if (node === null) return;
        if (this.currentSize === 1) this.head = null;
        else {
            if (node === this.head) this.head = node.next;
            node.previous!.next = node.next;
            node.next!.previous = node.previous;
        }
        this.window = null;
        this.currentSize--;
        this.invalidateAccessMemo();
    }
    public reverse(): void {
        if (this.head === null || this.currentSize < 2) return;
        let left = this.head;
        let right = this.head.previous!;
        for (let i = 0; i < Math.floor(this.currentSize / 2); i++) {
            const value = left.data;
            left.data = right.data;
            right.data = value;
            left = left.next!;
            right = right.previous!;
        }
        this.invalidateAccessMemo();
    }

    private insertAtEnd(element: E): void {
        const node = new _CircularDoubleLinkedListNode<E>();
        node.data = element;
        if (this.head === null) {
            node.next = node;
            node.previous = node;
            this.head = node;
        } else {
            const tail = this.head.previous!;
            node.previous = tail;
            node.next = this.head;
            tail.next = node;
            this.head.previous = node;
        }
        this.currentSize++;
        this.invalidateAccessMemo();
    }
    private insertBeforeHead(element: E): void {
        const node = new _CircularDoubleLinkedListNode<E>();
        node.data = element;
        if (this.head === null) {
            node.next = node;
            node.previous = node;
        } else {
            const tail = this.head.previous!;
            node.previous = tail;
            node.next = this.head;
            tail.next = node;
            this.head.previous = node;
        }
        this.head = node;
        this.currentSize++;
        this.invalidateAccessMemo();
    }
    private insertBeforeNode(element: E, pivot: _CircularDoubleLinkedListNode<E>): void {
        const node = new _CircularDoubleLinkedListNode<E>();
        node.data = element;
        node.previous = pivot.previous;
        node.next = pivot;
        pivot.previous!.next = node;
        pivot.previous = node;
        this.currentSize++;
        this.invalidateAccessMemo();
    }
    private nodeFor(element: E): _CircularDoubleLinkedListNode<E> | null {
        if (this.head === null) return null;
        let node = this.head;
        for (let i = 0; i < this.currentSize; i++, node = node.next!) if (node.data === element) return node;
        return null;
    }
    private invalidateAccessMemo(): void {
        this.accessMemoIndex = -1;
        this.accessMemoNode = null;
    }
}
